### PR TITLE
fix(helm): handle OCI version constraints and metadata

### PR DESCRIPTION
**Pull request by hiddeco**
Fixes the second bit of https://github.com/akuity/kargo/issues/2831#issuecomment-2435595674

---

When working with OCI-based Helm charts, version handling needs to be
stricter than with classic HTTP repositories:

1. OCI chart versions must be strict SemVer (e.g. "1.0.0" not "v1.0")
2. OCI registry tags cannot contain "+" characters used in SemVer
   build metadata

To solve this:

- Add strict SemVer validation for OCI chart versions
- Convert between OCI's "_" and SemVer's "+" for build metadata

### CODE REVIEW

File: internal/helm/helm.go [hunk 1]
```diff
@@ -185,10 +183,25 @@ func getChartVersionsFromOCIRepo(
 
 // versionsToSemVerCollection converts a slice of versions to a semver.Collection.
 // Any versions that cannot be parsed as SemVer are ignored.
-func versionsToSemVerCollection(versions []string) semver.Collection {
+func versionsToSemVerCollection(versions []string, isOCI bool) semver.Collection {
+	newSemver := semver.NewVersion
+	if isOCI {
+		// Image tags produced by Helm are STRICT SemVer, meaning that they
```
Comment (krancour) [inline internal/helm/helm.go:189]: Nit: Do we really want to say "Image tags" here? Applies in a few other places as well.

Totally non-blocking.

File: internal/helm/helm.go [hunk 2]
```diff
@@ -223,84 +242,6 @@ func filterSemVers(semvers semver.Collection, semverConstraint string) (semver.C
 	return filtered, nil
 }
 
-// Login runs `helm registry login` or `helm repo add` for the provided
-// repository. The provided homePath is used to set the HOME environment
-// variable, as well as the XDG_* environment variables. This ensures that Helm
-// uses the provided homePath as its configuration directory, and allows for
-// isolation.
-func Login(homePath, repository string, credentials Credentials) error {
```
Comment (krancour) [inline internal/helm/helm.go:231]: Dead code. Nice find.

**Review by krancour — APPROVED** [review]
Presuming you checked this out e2e, LGTM.

File: internal/helm/helm.go (hunk 1 above)
Comment (hiddeco) [reply internal/helm/helm.go]: Changed it to "OCI artifact tags".

**Comment by hiddeco** [comment]
@krancour I now thoroughly tested it end-to-end, and for our use case, it is better to stay true to the "Helm version". Reason being that ArgoCD and other bits where we utilize the version want the proper SemVer version, and not the registry one.

**Comment by krancour** [comment]
I'm not sure I followed that. Could you give an example?

**Comment by hiddeco** [comment]
Assuming there is a chart with `1.0.0+build.1`, this would get pushed to an OCI registry as `1.0.0_build.1`.

When the Warehouse is gathering chart versions, it now has two options:

1. Store the version reference to equal the tag.
2. Store the version reference to equal the strict SemVer.

The problem, if you go with option 1, is that if Freight is created with this and later consumed by a Promotion, this format is not valid to e.g. ArgoCD or any other consumer, as Helm always expects to receive strict SemVers when e.g. doing `helm pull`, etc.

Given this, it is better to store the OCI tag as a proper SemVer.

**Comment by krancour** [comment]
Ok. Then I understood it already after all. This appears to be what you're doing? I think I was confused because I'd somehow misinterpreted your previous comment as second-guessing this change.

**Comment by hiddeco** [comment]
I was not! But my initial  (now overwritten) approach was doing option 2. We are all good here! 😄

**Pull request merged**